openmc/arithmetic.py

```diff
@@ -0,0 +1,842 @@
+        return self._right_nuclide
+
+    @property
+    def binary_op(self):
+        return self._binary_op
+
+    @property
+    def name(self):
+
+        string = ''
+
+        # If the Summary was linked, the left nuclide is a Nuclide object
+        if isinstance(self.left_nuclide, openmc.Nuclide):
+            string += '(' + self.left_nuclide.name
+        # If the Summary was not linked, the left nuclide is the ZAID
+        else:
+            string += '(' + str(self.left_nuclide)
+
+        string += ' ' + self.binary_op + ' '
+
+        # If the Summary was linked, the right nuclide is a Nuclide object
+        if isinstance(self.right_nuclide, openmc.Nuclide):
+            string += self.right_nuclide.name + ')'
+        # If the Summary was not linked, the right nuclide is the ZAID
+        else:
+            string += str(self.right_nuclide) + ')'
+
+        return string
+
+    @left_nuclide.setter
+    def left_nuclide(self, left_nuclide):
+        cv.check_type('left_nuclide', left_nuclide,
+                      (openmc.Nuclide, CrossNuclide, AggregateNuclide))
+        self._left_nuclide = left_nuclide
+
+    @right_nuclide.setter
+    def right_nuclide(self, right_nuclide):
+        cv.check_type('right_nuclide', right_nuclide,
+                      (openmc.Nuclide, CrossNuclide, AggregateNuclide))
+        self._right_nuclide = right_nuclide
+
+    @binary_op.setter
+    def binary_op(self, binary_op):
+        cv.check_type('binary_op', binary_op, str)
+        cv.check_value('binary_op', binary_op, _TALLY_ARITHMETIC_OPS)
+        self._binary_op = binary_op
+
+
+class CrossFilter(object):
+    """A special-purpose filter used to encapsulate all combinations of two
+    tally's filter bins as an outer product for tally arithmetic.
+
+    Parameters
+    ----------
+    left_filter : Filter or CrossFilter
+        The left filter in the outer product
+    right_filter : Filter or CrossFilter
+        The right filter in the outer product
+    binary_op : str
+        The tally arithmetic binary operator (e.g., '+', '-', etc.) used to
+        combine two tally's filter bins with this CrossFilter
+
+    Attributes
+    ----------
+    type : str
+        The type of the crossfilter (e.g., 'energy / energy')
+    left_filter : Filter or CrossFilter
+        The left filter in the outer product
+    right_filter : Filter or CrossFilter
+        The right filter in the outer product
+    binary_op : str
+        The tally arithmetic binary operator (e.g., '+', '-', etc.) used to
+        combine two tally's filter bins with this CrossFilter
+    bins : dict of Iterable
+        A dictionary of the bins from each filter keyed by the types of the
+        left / right filters
+    num_bins : Integral
+        The number of filter bins (always 1 if aggregate_filter is defined)
+
+    """
+
+    def __init__(self, left_filter=None, right_filter=None, binary_op=None):
+
+        left_type = left_filter.type
```

Comment:
Seems like we should add some error checking here. If `left_filter` is None (it is defined None as a default argument so I assume this is a possibility), then you will get an AttributeError exception raised.